client/src/main/java/io/pravega/client/batch/impl/BatchClientImpl.java

```diff
@@ -0,0 +1,100 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.client.batch.impl;
+
+import com.google.common.annotations.Beta;
+import com.google.common.collect.Iterators;
+import io.pravega.client.batch.BatchClient;
+import io.pravega.client.batch.SegmentInfo;
+import io.pravega.client.batch.SegmentIterator;
+import io.pravega.client.batch.StreamInfo;
+import io.pravega.client.netty.impl.ConnectionFactory;
+import io.pravega.client.segment.impl.Segment;
+import io.pravega.client.segment.impl.SegmentInputStreamFactory;
+import io.pravega.client.segment.impl.SegmentInputStreamFactoryImpl;
+import io.pravega.client.segment.impl.SegmentMetadataClient;
+import io.pravega.client.segment.impl.SegmentMetadataClientFactory;
+import io.pravega.client.segment.impl.SegmentMetadataClientFactoryImpl;
+import io.pravega.client.stream.Serializer;
+import io.pravega.client.stream.Stream;
+import io.pravega.client.stream.impl.Controller;
+import io.pravega.client.stream.impl.StreamCut;
+import io.pravega.client.stream.impl.StreamImpl;
+import java.util.Date;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import lombok.Cleanup;
+import org.apache.commons.lang3.NotImplementedException;
+
+import static io.pravega.common.concurrent.FutureHelpers.getAndHandleExceptions;
+
+@Beta
+public class BatchClientImpl implements BatchClient {
+
+    private final Controller controller;
+    private final SegmentInputStreamFactory inputStreamFactory;
+    private final SegmentMetadataClientFactory segmentMetadataClientFactory;
+
+    public BatchClientImpl(Controller controller, ConnectionFactory connectionFactory) {
+        this.controller = controller;
+        inputStreamFactory = new SegmentInputStreamFactoryImpl(controller, connectionFactory);
+        segmentMetadataClientFactory = new SegmentMetadataClientFactoryImpl(controller, connectionFactory);
+    }
+    
+    private StreamInfo getStreamInfo(Stream stream) {
+        // TODO: Implement this method and make it public
+        // Name from stream
+        // Length refector from ReaderGroupImpl perhaps move to controller.
+        // Creation time needs an added api? or perhaps modify the getsegmentAtTime api
+        // create a controller.getStreamSealTime() which returns null if open
+        throw new NotImplementedException("getStreamInfo");
+    }
+
+    @Override
+    public Iterator<SegmentInfo> listSegments(Stream stream) {
+        return listSegments(stream, new Date(0L));
+    }
+
+    private Iterator<SegmentInfo> listSegments(Stream stream, Date from) {
```

Comment:
shall we also add "listSegments(Stream stream, Date from)" to batchclient interface?

Reply:
Actually, it is the other way around, we have chosen to remove the list segments based on a from date:

https://github.com/pravega/pravega/pull/1998#discussion_r146007569

we don't need this argument here any longer, @tkaitchuck probably forgot to remove it.

Reply:
it means hadoop jobs need to process all events from very beginning, but not from the first segment containing the events created after the specified time. Does it meet customer's expectation? It's OK if decision has been made, just want to make sure our output is exactly what customer wanted.

Reply:
Thanks for the input, @yangb8 .

The notion of time we have in the controller today is coarse and makes very weak guarantees about the distance from the reference to the actual point it starts at, so even if we add it now, the results would be underwhelming.

I'm +1 for not mixing up issues. My preference is that this first cut of the batch read API has no notion of partial reads of a stream, and as such, I suggest not to include any reference, time-based oar not, to arbitrary points of a stream to limit the processing of data. We have the notion of stream cuts that we are promoting as references to arbitrary points of a stream and incorporating it to this API is subject of future work.

Reply:
Thanks for the explanation in details!